adapter/pkg/oasparser/envoyconf/routesWithClusters.go

```diff
@@ -48,117 +49,138 @@ import (
-		apilevelClusterSand = createCluster(apilevelAddressSand, apiLevelClusterNameSand, apiLevelEndpointSand[0].URLType)
-		clustersSand = append(clustersSand, apilevelClusterSand)
-		endpointsSand = append(endpointsSand, apilevelAddressSand)
-	}
+	// To keep track of API Level production endpoint basePath
+	apiEndpointBasePath := ""
+
+	apiTitle := mgwSwagger.GetTitle()
+	apiVersion := mgwSwagger.GetVersion()
+	apiBasePath := mgwSwagger.GetXWso2Basepath()
 
 	// check API level production endpoints available
 	if len(mgwSwagger.GetProdEndpoints()) > 0 {
 		apiLevelEndpointProd = mgwSwagger.GetProdEndpoints()
 		apilevelAddressP := createAddress(apiLevelEndpointProd[0].Host, apiLevelEndpointProd[0].Port)
 		apiLevelClusterNameProd = strings.TrimSpace(prodClustersConfigNamePrefix +
 			strings.Replace(mgwSwagger.GetTitle(), " ", "", -1) + mgwSwagger.GetVersion())
-		apilevelClusterProd = createCluster(apilevelAddressP, apiLevelClusterNameProd, apiLevelEndpointProd[0].URLType)
-		clustersProd = append(clustersProd, apilevelClusterProd)
-		endpointsProd = append(endpointsProd, apilevelAddressP)
+		apilevelClusterProd = createCluster(apilevelAddressP, apiLevelClusterNameProd, apiLevelEndpointProd[0].URLType,
+			upstreamCerts)
+		clusters = append(clusters, apilevelClusterProd)
+		endpoints = append(endpoints, apilevelAddressP)
+		apiEndpointBasePath = apiLevelEndpointProd[0].Basepath
 
 	} else {
 		logger.LoggerOasparser.Warn("API level Producton endpoints are not defined")
 	}
-	for _, resource := range mgwSwagger.GetResources() {
-		apiTitle := mgwSwagger.GetTitle()
-		apiVersion := mgwSwagger.GetVersion()
-		apiBasePath := mgwSwagger.GetXWso2Basepath()
 
-		// resource level check sandbox endpoints
-		if len(resource.GetSandEndpoints()) > 0 {
-			endpointSand = resource.GetSandEndpoints()
-			addressSand := createAddress(endpointSand[0].Host, endpointSand[0].Port)
-			// TODO: (VirajSalaka) 0 is hardcoded as only one endpoint is supported at the moment
-			clusterNameSand := strings.TrimSpace(apiLevelClusterNameSand + "_" + strings.Replace(resource.GetID(), " ", "", -1) +
-				"0")
-			clusterSand := createCluster(addressSand, clusterNameSand, endpointSand[0].URLType)
-			clustersSand = append(clustersSand, clusterSand)
-			clusterRefSand := clusterSand.GetName()
-
-			// sandbox endpoints
-			routeS := createRoute(apiTitle, apiBasePath, apiVersion, endpointSand[0], resource, clusterRefSand)
-			routesSand = append(routesSand, routeS)
-			endpointsSand = append(endpointsSand, addressSand)
-
-			// API level check
-		} else if len(mgwSwagger.GetSandEndpoints()) > 0 {
-			endpointSand = apiLevelEndpointSand
-			clusterRefSand := apilevelClusterSand.GetName()
+	// check API level sandbox endpoints availble
+	if len(mgwSwagger.GetSandEndpoints()) > 0 {
+		apiLevelEndpointSand = mgwSwagger.GetSandEndpoints()
+		if apiEndpointBasePath != apiLevelEndpointSand[0].Basepath && len(apiLevelEndpointProd) > 0 {
```

Comment:
within Ext authz, do path rewrite

Reply:
This approach has not been successful. When it is invoked, the client will always get a 404. The following log corresponds to a similar behavior where petstore v2 /pet/findbyStatus resource is called and the ext_authz filter changed that path to some random string (/aaaaa/vvvvv)

```
router_1    | [2021-01-05 20:24:10.414][18][trace][filter] [source/extensions/filters/http/ext_authz/ext_authz.cc:69] [C1][S4670466387945752909] ext_authz filter calling authorization server
router_1    | [2021-01-05 20:24:10.415][18][trace][ext_authz] [source/extensions/filters/common/ext_authz/ext_authz_grpc_impl.cc:60] Sending CheckRequest: attributes {
router_1    |   source {
router_1    |     address {
router_1    |       socket_address {
router_1    |         address: "172.18.0.1"
router_1    |         port_value: 48154
router_1    |       }
router_1    |     }
router_1    |   }
router_1    |   destination {
router_1    |     address {
router_1    |       socket_address {
router_1    |         address: "172.18.0.4"
router_1    |         port_value: 9095
router_1    |       }
router_1    |     }
router_1    |     principal: "localhost"
router_1    |   }
router_1    |   request {
router_1    |     time {
router_1    |       seconds: 1609878250
router_1    |       nanos: 414538000
router_1    |     }
router_1    |     http {
router_1    |       id: "4670466387945752909"
router_1    |       method: "GET"
router_1    |       headers {
router_1    |         key: ":authority"
router_1    |         value: "localhost:9095"
router_1    |       }
router_1    |       headers {
router_1    |         key: ":method"
router_1    |         value: "GET"
router_1    |       }
router_1    |       headers {
router_1    |         key: ":path"
router_1    |         value: "/v2/pet/findByStatus?status=available"
router_1    |       }
router_1    |       headers {
router_1    |         key: "accept"
router_1    |         value: "application/xml"
router_1    |       }
router_1    |       headers {
router_1    |         key: "authorization"
router_1    |         value: "bearer eyJ4NXQiOiJNell4TW1Ga09HWXdNV0kwWldObU5EY3hOR1l3WW1NNFpUQTNNV0kyTkRBelpHUXpOR00wWkdSbE5qSmtPREZrWkRSaU9URmtNV0ZoTXpVMlpHVmxOZyIsImtpZCI6Ik16WXhNbUZrT0dZd01XSTBaV05tTkRjeE5HWXdZbU00WlRBM01XSTJOREF6WkdRek5HTTBaR1JsTmpKa09ERmtaRFJpT1RGa01XRmhNelUyWkdWbE5nX1JTMjU2IiwiYWxnIjoiUlMyNTYifQ==.eyJhdWQiOiJBT2syNFF6WndRXzYyb2QyNDdXQnVtd0VFZndhIiwic3ViIjoiYWRtaW5AY2FyYm9uLnN1cGVyIiwibmJmIjoxNTk2MDA5NTU2LCJhenAiOiJBT2syNFF6WndRXzYyb2QyNDdXQnVtd0VFZndhIiwic2NvcGUiOiJhbV9hcHBsaWNhdGlvbl9zY29wZSBkZWZhdWx0IiwiaXNzIjoiaHR0cHM6Ly9sb2NhbGhvc3Q6OTQ0My9vYXV0aDIvdG9rZW4iLCJrZXl0eXBlIjoiUFJPRFVDVElPTiIsImV4cCI6MTYyNzU0NTU1NiwiaWF0IjoxNTk2MDA5NTU2LCJqdGkiOiIyN2ZkMWY4Ny01ZTI1LTQ1NjktYTJkYi04MDA3MTFlZTJjZWMifQ==.otDREOsUUmXuSbIVII7FR59HAWqtXh6WWCSX6NDylVIFfED3GbLkopo6rwCh2EX6yiP-vGTqX8sB9Zfn784cIfD3jz2hCZqOqNzSUrzamZrWui4hlYC6qt4YviMbR9LNtxxu7uQD7QMbpZQiJ5owslaASWQvFTJgBmss5t7cnurrfkatj5AkzVdKOTGxcZZPX8WrV_Mo2-rLbYMslgb2jCptgvi29VMPo9GlAFecoMsSwywL8sMyf7AJ3y4XW5Uzq7vDGxojDam7jI5W8uLVVolZPDstqqZYzxpPJ2hBFC_OZgWG3LqhUgsYNReDKKeWUIEieK7QPgjetOZ5Geb1mA=="
router_1    |       }
router_1    |       headers {
router_1    |         key: "user-agent"
router_1    |         value: "curl/7.64.1"
router_1    |       }
router_1    |       headers {
router_1    |         key: "x-forwarded-proto"
router_1    |         value: "https"
router_1    |       }
router_1    |       headers {
router_1    |         key: "x-request-id"
router_1    |         value: "0b7dbc1e-d59c-49ee-a472-b8498cc4e285"
router_1    |       }
router_1    |       path: "/v2/pet/findByStatus?status=available"
router_1    |       host: "localhost:9095"
router_1    |       protocol: "HTTP/1.1"
router_1    |     }
router_1    |   }
router_1    |   context_extensions {
router_1    |     key: "basePath"
router_1    |     value: "/v2"
router_1    |   }
router_1    |   context_extensions {
router_1    |     key: "method"
router_1    |     value: "GET"
router_1    |   }
router_1    |   context_extensions {
router_1    |     key: "name"
router_1    |     value: "Swagger Petstore"
router_1    |   }
router_1    |   context_extensions {
router_1    |     key: "path"
router_1    |     value: "/pet/findByStatus"
router_1    |   }
router_1    |   context_extensions {
router_1    |     key: "prodClusterName"
router_1    |     value: "clusterProd_SwaggerPetstore1.0.5"
router_1    |   }
router_1    |   context_extensions {
router_1    |     key: "sandClusterName"
router_1    |     value: "clusterSand_SwaggerPetstore1.0.5"
router_1    |   }
router_1    |   context_extensions {
router_1    |     key: "version"
router_1    |     value: "1.0.5"
router_1    |   }
router_1    |   metadata_context {
router_1    |   }
router_1    | }
router_1    | 
router_1    | [2021-01-05 20:24:10.415][18][debug][router] [source/common/router/router.cc:429] [C0][S13498356977599373084] cluster 'ext-authz' match for URL '/envoy.service.auth.v3.Authorization/Check'
router_1    | [2021-01-05 20:24:10.415][18][debug][router] [source/common/router/router.cc:586] [C0][S13498356977599373084] router decoding headers:
router_1    | ':method', 'POST'
router_1    | ':path', '/envoy.service.auth.v3.Authorization/Check'
router_1    | ':authority', 'ext-authz'
router_1    | ':scheme', 'http'
router_1    | 'te', 'trailers'
router_1    | 'content-type', 'application/grpc'
router_1    | 'x-envoy-internal', 'true'
router_1    | 'x-forwarded-for', '172.18.0.4'
router_1    | 
router_1    | [2021-01-05 20:24:10.415][18][trace][router] [source/common/router/upstream_request.cc:211] [C0][S13498356977599373084] buffering 1680 bytes
router_1    | [2021-01-05 20:24:10.416][18][debug][router] [source/common/router/upstream_request.cc:357] [C0][S13498356977599373084] pool ready
router_1    | [2021-01-05 20:24:10.945][18][debug][router] [source/common/router/router.cc:1178] [C0][S13498356977599373084] upstream headers complete: end_stream=false
router_1    | [2021-01-05 20:24:10.946][18][trace][ext_authz] [source/extensions/filters/common/ext_authz/ext_authz_grpc_impl.cc:67] Received CheckResponse: status {
router_1    | }
router_1    | ok_response {
router_1    |   headers {
router_1    |     header {
router_1    |       key: "x-wso2-cluster-header"
router_1    |       value: "clusterProd_SwaggerPetstore1.0.5"
router_1    |     }
router_1    |   }
router_1    |   headers {
router_1    |     header {
router_1    |       key: ":path"
router_1    |       value: "/aaaaa/vvvvv"
router_1    |     }
router_1    |   }
router_1    | }
router_1    | 
router_1    | [2021-01-05 20:24:10.946][18][debug][filter] [source/extensions/filters/http/ext_authz/ext_authz.cc:180] [C1][S4670466387945752909] ext_authz is clearing route cache
router_1    | [2021-01-05 20:24:10.946][18][trace][filter] [source/extensions/filters/http/ext_authz/ext_authz.cc:184] [C1][S4670466387945752909] ext_authz filter added header(s) to the request:
router_1    | [2021-01-05 20:24:10.946][18][trace][filter] [source/extensions/filters/http/ext_authz/ext_authz.cc:186] [C1][S4670466387945752909] 'x-wso2-cluster-header':'clusterProd_SwaggerPetstore1.0.5'
router_1    | [2021-01-05 20:24:10.946][18][trace][filter] [source/extensions/filters/http/ext_authz/ext_authz.cc:186] [C1][S4670466387945752909] ':path':'/aaaaa/vvvvv'
router_1    | [2021-01-05 20:24:10.946][18][trace][filter] [source/extensions/filters/http/ext_authz/ext_authz.cc:212] [C1][S4670466387945752909] ext_authz filter removed header(s) from the request:
router_1    | [2021-01-05 20:24:10.948][18][debug][router] [source/common/router/router.cc:364] [C1][S4670466387945752909] no cluster match for URL '/aaaaa/vvvvv'
```